contracts/src/v0.8/upkeeps/ERC20BalanceMonitor.sol

```diff
@@ -0,0 +1,289 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.4;
+
+import "../ConfirmedOwner.sol";
+import "../interfaces/KeeperCompatibleInterface.sol";
+import "@openzeppelin/contracts/security/Pausable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+/**
+ * @title The ERC20BalanceMonitor contract.
+ * @notice A keeper-compatible contract that monitors and funds ERC20 tokens.
+ */
+contract ERC20BalanceMonitor is ConfirmedOwner, Pausable, KeeperCompatibleInterface {
+  IERC20 private ERC20Token;
+
+  uint256 private constant MIN_GAS_FOR_TRANSFER = 55_000;
+
+  event FundsAdded(uint256 amountAdded, uint256 newBalance, address sender);
+  event FundsWithdrawn(uint256 amountWithdrawn, address payee);
+  event TopUpSucceeded(address indexed topUpAddress);
+  event TopUpFailed(address indexed topUpAddress);
+  event WatchlistUpdated(address[] oldWatchlist, address[] newWatchlist);
+  event KeeperRegistryAddressUpdated(address oldAddress, address newAddress);
+  event ERC20TokenAddressUpdated(address oldAddress, address newAddress);
+  event MinWaitPeriodUpdated(uint256 oldMinWaitPeriod, uint256 newMinWaitPeriod);
+
+  error InvalidWatchList();
+  error OnlyKeeperRegistry();
+  error DuplicateAddress(address duplicate);
+
+  struct Target {
+    bool isActive;
+    uint96 minBalance;
+    uint96 topUpAmount;
+    uint56 lastTopUpTimestamp;
+  }
+
+  address private s_keeperRegistryAddress;
+  uint256 private s_minWaitPeriodSeconds;
+  address[] private s_watchList;
+  mapping(address => Target) internal s_targets;
+
+  /**
+   * @param erc20TokenAddress the ERC20 token address
+   * @param keeperRegistryAddress the address of the keeper registry contract
+   * @param minWaitPeriodSeconds the minimum wait period for addresses between funding
+   */
+  constructor(
+    address erc20TokenAddress,
+    address keeperRegistryAddress,
+    uint256 minWaitPeriodSeconds
+  ) ConfirmedOwner(msg.sender) {
+    setERC20TokenAddress(erc20TokenAddress);
+    setKeeperRegistryAddress(keeperRegistryAddress);
+    setMinWaitPeriodSeconds(minWaitPeriodSeconds);
+  }
+
+  /**
+   * @notice Sets the list of subscriptions to watch and their funding parameters.
+   * @param addresses the list of subscription ids to watch
+   * @param minBalances the minimum balances for each subscription
+   * @param topUpAmounts the amount to top up each subscription
+   */
+  function setWatchList(
+    address[] calldata addresses,
+    uint96[] calldata minBalances,
+    uint96[] calldata topUpAmounts
+  ) external onlyOwner {
+    if (addresses.length != minBalances.length || addresses.length != topUpAmounts.length) {
+      revert InvalidWatchList();
+    }
```

Comment:
might be good to have some limits on number of addresses allowed, o/w checkUpkeep can run OOG

Reply:
Any guidance on a max here? I'm thinking of a relatively large hardcoded value like 1000.

Reply:
checkUpkeep gas limit is 6.5M today, 300 should fit into that but just verify once